From 01b35aa0883ab4aa6a0929a88b162080b0294ed2 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Thu, 3 Jan 2019 21:20:42 +0100
Subject: [PATCH 2/3] update_engine: Transition to backuptool domain

* This way we can make backuptool permissive
  while leaving update_engine rules as clean as possible

Change-Id: Ied17f31f7d0258319371f3badc891c7a6ecee901
---
 Android.mk                                    |  1 +
 payload_consumer/postinstall_runner_action.cc | 13 +++++++++++++
 2 files changed, 14 insertions(+)

diff --git a/Android.mk b/Android.mk
index 8f2c8fa..35f8f52 100644
--- a/Android.mk
+++ b/Android.mk
@@ -110,6 +110,7 @@ ue_libpayload_consumer_exported_static_libraries := \
     $(ue_update_metadata_protos_exported_static_libraries)
 ue_libpayload_consumer_exported_shared_libraries := \
     libcrypto \
+    libselinux \
     $(ue_update_metadata_protos_exported_shared_libraries)
 
 ue_libpayload_consumer_src_files := \
diff --git a/payload_consumer/postinstall_runner_action.cc b/payload_consumer/postinstall_runner_action.cc
index acdf9d7..1ba6db1 100644
--- a/payload_consumer/postinstall_runner_action.cc
+++ b/payload_consumer/postinstall_runner_action.cc
@@ -19,6 +19,7 @@
 #include <fcntl.h>
 #include <signal.h>
 #include <stdlib.h>
+#include <selinux/selinux.h>
 #include <sys/mount.h>
 #include <sys/types.h>
 #include <unistd.h>
@@ -179,11 +180,23 @@ void PostinstallRunnerAction::PerformPartitionPostinstall() {
     utils::MountFilesystem(mountable_device, fs_mount_dir_, MS_NOATIME | MS_NODEV | MS_NODIRATIME,
                            partition.filesystem_type, "seclabel");
 
+    // Switch to a permissive domain
+    if (setexeccon("u:r:backuptool:s0")) {
+      LOG(ERROR) << "Failed to set backuptool context";
+      return CompletePostinstall(ErrorCode::kPostinstallRunnerError);
+    }
+
     // Run backuptool script
     int ret = system("/postinstall/system/bin/backuptool_postinstall.sh");
     if (ret == -1 || WEXITSTATUS(ret) != 0) {
       LOG(ERROR) << "Backuptool postinstall step failed. ret=" << ret;
     }
+
+    // Switch back to update_engine domain
+    if (setexeccon(nullptr)) {
+      LOG(ERROR) << "Failed to set update_engine context";
+      return CompletePostinstall(ErrorCode::kPostinstallRunnerError);
+    }
   } else {
     LOG(INFO) << "Skipping backuptool scripts";
   }
-- 
2.17.1

